notebooks/ntsb-duckdb.ipynb

```diff
@@ -0,0 +1,340 @@
+    "first_two = table_data[2:4]\n",
+    "\n",
+    "print(\"Text representation of the first two values (not including headers):\")\n",
+    "for i, row in enumerate(first_two, 1):\n",
+    "    print(f\"\\n{i}. doc_id: {row[0]}\")\n",
+    "    print(f\"   Text representation:\\n   {row[1]}\")\n",
+    "    print(\"-\" * 50)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# We can also see the performance of the DuckDB query and check the index is being used by using the in-built visualization below \n",
+    "q = \"What automobile type is the most accident prone?\"\n",
+    "explain_result = in_memory_db.sql(f\"EXPLAIN SELECT doc_id, text_representation, metadata FROM in_memory_table ORDER BY array_distance(embeddings, {embedder.embed_query(q)}::FLOAT[384]) LIMIT 10\")\n",
+    "\n",
+    "# Convert the result to a list of lists\n",
+    "explain_data = explain_result.fetchall()\n",
+    "\n",
+    "# Get column names\n",
+    "headers = [desc[0] for desc in explain_result.description]\n",
+    "\n",
+    "# Pretty print the EXPLAIN output\n",
+    "print(\"Expanded EXPLAIN output:\")\n",
+    "print(tabulate(explain_data, headers=headers, tablefmt=\"grid\"))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# We now initialize a vector store on DuckDB using the LangChain integration. We perform similarity search using the integration here\n",
+    "from langchain_community.vectorstores.duckdb import DuckDB\n",
+    "vector_store = DuckDB(connection=in_memory_db, embedding=embedder, id_key= \"doc_id\", text_key=\"text_representation\", vector_key=\"embeddings\", table_name=\"in_memory_table\")\n",
+    "result = vector_store.similarity_search('traffic')\n",
+    "print(result)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from langchain_openai import ChatOpenAI  \n",
+    "from langchain.chains import RetrievalQAWithSourcesChain  \n",
+    "import os\n",
+    "# Finally, we initialize a RAG agent and ask a model question of the data\n",
+    "llm = ChatOpenAI(  \n",
+    "    openai_api_key=os.environ.get(\"OPENAI_API_KEY\"),  \n",
+    "    model_name='gpt-3.5-turbo',  \n",
+    "    temperature=0.8  \n",
+    ")  \n",
+    "qa = RetrievalQAWithSourcesChain.from_chain_type(  \n",
+    "    llm=llm,  \n",
+    "    chain_type=\"stuff\",  \n",
+    "    retriever=vector_store.as_retriever() , verbose=True\n",
+    ")  \n",
+    "qa.invoke({\"question\": \"How many accidents happened?\"})"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Finally we can build an index with other distance metrics, the default is Euclidean Distance ('l2sq'), with Cosine similarity ('cosine', 'array_cosine_simarlity') \n",
```

Comment:
We aren't really explaining why you'd wand to do this. I might phrase it as "To experiment with XXXXX, we can..."